docs/src/proposals/blockstore-rocksdb-compaction.md

```diff
@@ -0,0 +1,234 @@
+required for storing each entry.  In other words, RocksDB uses compactions
+to balance [write, space, and read amplifications](https://smalldatum.blogspot.com/2015/11/read-write-space-amplification-pick-2_23.html).
+
+As different workloads have different requirements, RocksDB makes its options
+highly configerable.  However, it also means its default settings might not
+be always suitable.  This document proposes some directions specifically
+optimized for Solana's use case.
+
+## Problems
+As mentioned in [#16234](https://github.com/solana-labs/solana/issues/16234),
+there're several issues in the Solana's BlockStore which runs RocksDB with
+level compaction.  Here's a quick summary of the issues:
+
+### Long Write Stalls
+Remember that RocksDB periodically runs background compactions in order to
+keep the number of logical levels small in order to reach the target read
+amplification.  However, when the background compactions cannot catch up
+the write rate, the number of logical levels will eventually exceeds the
+configured limit.  In such case, RocksDB will rate-limit / stop all writes
+when it reaches soft / hard threshold.
+
+In [#14586](https://github.com/solana-labs/solana/issues/14586), it is reported
+that the write stalls in Solana's use case can be 40 minutes long.  It is also
+reported in [#16234](https://github.com/solana-labs/solana/issues/16234) that
+writes are also slowed-down, indicating the underlying RocksDB instance has
+reach the soft limit for write stall.
+
+### Delections are not Processed in Time
+Deletions are processed in the same way as other write operations in RocksDB,
+where multiple entries are merged / deleted during the background compaction.
+Although the deleted entries will not be visible from the read side right
+after the deletion is issued, the deleted entries (including the original
+data entries and its deletion entries) will still occupy disk storage.
+
+TBD: explain how write-key order makes this worse
+
+### High Write I/O from Write Amplification
+In addition to write stalls, it is also observed that compactions cause
+unwanted high write I/O.  With the current design where level compaction
+is configured for BlockStore, it has ~30x write amplification (10x write amp
+per level and assuming three levels in average).
+
+## Current Design
+In the current BlockStore, The default level compaction + slot-ID based
+compaction filter + periodic compaction are used, which has ~30x write
+amplification (10x write amp per level and assuming three levels in
+average + the write amp caused by the periodic compaction.)
+
+## Proposed Design -- FIFO Compaction with TTL + Rate Limiting
+As all the above issues are compaction related, it can be solved with
+a proper compaction style and deletion settings.  Fortunately, Solana's
+BlockStore has an unique workload where write-keys are mostly monotonically
+increasing over time.  This allows data to be persisted in sorted order
```

Comment:
well, when repairing, this might not true. that's why I avoided fifo at the time.